server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Mailbox.scala

```diff
@@ -0,0 +1,107 @@
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ * http://www.apache.org/licenses/LICENSE-2.0                   *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+
+package org.apache.james.jmap.mail
+
+import org.apache.james.core.Username
+import org.apache.james.jmap.model.UnsignedInt
+import org.apache.james.mailbox.Role
+import org.apache.james.mailbox.model.MailboxId
+
+final case class MailboxName(name: String) {
+  require(name != null, "'name' is mandatory")
+  require(!name.isEmpty, "'name' is mandatory")
+}
+
+final case class MailboxRights(mayReadItems: Boolean,
+                               mayAddItems: Boolean,
+                               mayRemoveItems: Boolean,
+                               maySetSeen: Boolean,
+                               maySetKeywords: Boolean,
+                               mayCreateChild: Boolean,
+                               mayRename: Boolean,
+                               mayDelete: Boolean,
+                               maySubmit: Boolean)
+
+object MailboxNamespace {
+  sealed case class Type(value: String)
+
+  val Delegated = Type("Delegated")
+  val Personal = Type("Personal")
+
+  def delegated(owner: Username): MailboxNamespace = {
+    require(owner != null)
+    require(!owner.asString.trim.isEmpty)
+
+    new MailboxNamespace(Delegated, Some(owner))
+  }
+
+  def personal = new MailboxNamespace(Personal, None)
+}
+
+case class MailboxNamespace private(`type`: MailboxNamespace.Type, owner: Option[Username])
+
+object SortOrder {
+  private val DEFAULT_SORT_ORDER = SortOrder.of(1000L)
+  private val defaultSortOrders = Map(
+    Role.INBOX -> SortOrder.of(10L),
+    Role.ARCHIVE -> SortOrder.of(20L),
+    Role.DRAFTS -> SortOrder.of(30L),
+    Role.OUTBOX -> SortOrder.of(40L),
+    Role.SENT -> SortOrder.of(50L),
+    Role.TRASH -> SortOrder.of(60L),
+    Role.SPAM -> SortOrder.of(70L),
+    Role.TEMPLATES -> SortOrder.of(80L),
+    Role.RESTORED_MESSAGES -> SortOrder.of(90L))
+
+  def getSortOrder(role: Option[Role]): SortOrder = role.flatMap(r => getDefaultSortOrder(r))
+    .getOrElse(DEFAULT_SORT_ORDER)
+
+  private def getDefaultSortOrder(role: Role) = defaultSortOrders.get(role)
+
+  def of(sortOrder: UnsignedInt): SortOrder = {
```

Comment:
in scala, we usually rely on `apply` method for simple constructors. BTW, a case class provides that out of the box.